smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java

```diff
@@ -348,39 +346,24 @@ private MamQueryIQ prepareMamQueryIQGet(String queryId) {
         return mamQueryIQ;
     }
 
-    private MamQueryResult queryArchive(MamQueryIQ mamQueryIq, long extraTimeout) throws NoResponseException,
-            XMPPErrorException, NotConnectedException, InterruptedException, NotLoggedInException {
-        if (extraTimeout < 0) {
-            throw new IllegalArgumentException("extra timeout must be zero or positive");
-        }
-
+    private MamQueryResult queryArchive(MamQueryIQ mamQueryIq) throws NoResponseException, XMPPErrorException,
+            NotConnectedException, InterruptedException, NotLoggedInException {
         final XMPPConnection connection = getAuthenticatedConnectionOrThrow();
-        MamFinIQ mamFinIQ = null;
-
-        PacketCollector finIQCollector = connection.createPacketCollector(new MamFinIQFilter(mamQueryIq));
 
         PacketCollector.Configuration resultCollectorConfiguration = PacketCollector.newConfiguration()
-                .setStanzaFilter(new MamResultFilter(mamQueryIq)).setCollectorToReset(finIQCollector);
-
+                .setStanzaFilter(new MamResultFilter(mamQueryIq));
         PacketCollector resultCollector = connection.createPacketCollector(resultCollectorConfiguration);
 
-        try {
-            connection.createPacketCollectorAndSend(mamQueryIq).nextResultOrThrow();
-            IQ iq = finIQCollector.nextResultOrThrow(connection.getPacketReplyTimeout() + extraTimeout);
-            mamFinIQ = MamFinIQ.from(iq);
-        } finally {
-            resultCollector.cancel();
-            finIQCollector.cancel();
-        }
```

Comment:
Well it appears the code is missing now two critical non-functional properties:
1. Cancel both collectors in any case, in order to avoid leaking any collector
2. setCollectorToReset() is not used

Especially 2. is required in the retrieval pattern MAM (and similar XEPs) uses:
IQ-REQ → M1 → ... → MN → IQ-RES
The collector for the messages (M1-MN) needs to reset the timeout of the collector of the IQ response (IQ-RES), as a MAM (and similar) queries could potentially return a large data set, and thus you want to extend IQ collectors timeout over the normal IQ request-response timeout as long as data is arriving.